pkg/router/template/util/util.go

```diff
@@ -51,6 +51,12 @@ func GenerateRouteRegexp(hostname, path string, wildcard bool) string {
 		subpathRE = "(/.*)?"
 	}
 
+	// The path could contain space characters, which must be escaped in
+	// HAProxy map files.  See
+	// <https://bugzilla.redhat.com/show_bug.cgi?id=2074304>.
+	pathRE = strings.ReplaceAll(pathRE, " ", "\\ ")
+	pathRE = strings.ReplaceAll(pathRE, "\t", "\\\t")
```

Comment:
Do we also have to consider newlines?

Reply:
Newlines don't seem to cause a problem:
```
% oc -n openshift-ingress-canary create route edge test-route-with-path-with-cr --service=ingress-canary --path=$'/foo\rbar'
route.route.openshift.io/test-route-with-path-with-cr created
% oc -n openshift-ingress logs -c router -l ingresscontroller.operator.openshift.io/deployment-ingresscontroller=default --tail=5
I0412 19:04:47.854197       1 router.go:618] template "msg"="router reloaded"  "output"=" - Checking http://localhost:80 ...\n - Health check ok : 0 retry attempt(s).\n"
I0412 19:05:02.484520       1 router.go:618] template "msg"="router reloaded"  "output"=" - Checking http://localhost:80 ...\n - Health check ok : 0 retry attempt(s).\n"
I0412 19:06:08.034955       1 router.go:618] template "msg"="router reloaded"  "output"=" - Checking http://localhost:80 ...\n - Health check ok : 0 retry attempt(s).\n"
I0412 19:07:23.523944       1 router.go:618] template "msg"="router reloaded"  "output"=" - Checking http://localhost:80 ...\n - Health check ok : 0 retry attempt(s).\n"
I0412 19:13:26.570714       1 router.go:618] template "msg"="router reloaded"  "output"=" - Checking http://localhost:80 ...\n - Health check ok : 0 retry attempt(s).\n"
W0412 19:04:57.246318       1 reflector.go:442] github.com/openshift/router/pkg/router/controller/factory/factory.go:125: watch of *v1.Route ended with: an error on the server ("unable to decode an event from the watch stream: stream error: stream ID 7; INTERNAL_ERROR") has prevented the request from succeeding                                                                                                                
I0412 19:05:02.477983       1 router.go:618] template "msg"="router reloaded"  "output"=" - Checking http://localhost:80 ...\n - Health check ok : 0 retry attempt(s).\n"
I0412 19:06:08.034481       1 router.go:618] template "msg"="router reloaded"  "output"=" - Checking http://localhost:80 ...\n - Health check ok : 0 retry attempt(s).\n"
I0412 19:07:23.513991       1 router.go:618] template "msg"="router reloaded"  "output"=" - Checking http://localhost:80 ...\n - Health check ok : 0 retry attempt(s).\n"
I0412 19:13:26.571833       1 router.go:618] template "msg"="router reloaded"  "output"=" - Checking http://localhost:80 ...\n - Health check ok : 0 retry attempt(s).\n"
% oc -n openshift-ingress-canary create route edge test-route-with-path-with-ln --service=ingress-canary --path=$'/foo\nbar'
route.route.openshift.io/test-route-with-path-with-ln created
% oc -n openshift-ingress logs -c router -l ingresscontroller.operator.openshift.io/deployment-ingresscontroller=default --tail=5
I0412 19:05:02.484520       1 router.go:618] template "msg"="router reloaded"  "output"=" - Checking http://localhost:80 ...\n - Health check ok : 0 retry attempt(s).\n"
I0412 19:06:08.034955       1 router.go:618] template "msg"="router reloaded"  "output"=" - Checking http://localhost:80 ...\n - Health check ok : 0 retry attempt(s).\n"
I0412 19:07:23.523944       1 router.go:618] template "msg"="router reloaded"  "output"=" - Checking http://localhost:80 ...\n - Health check ok : 0 retry attempt(s).\n"
I0412 19:13:26.570714       1 router.go:618] template "msg"="router reloaded"  "output"=" - Checking http://localhost:80 ...\n - Health check ok : 0 retry attempt(s).\n"
I0412 19:13:36.540289       1 router.go:618] template "msg"="router reloaded"  "output"=" - Checking http://localhost:80 ...\n - Health check ok : 0 retry attempt(s).\n"
I0412 19:05:02.477983       1 router.go:618] template "msg"="router reloaded"  "output"=" - Checking http://localhost:80 ...\n - Health check ok : 0 retry attempt(s).\n"
I0412 19:06:08.034481       1 router.go:618] template "msg"="router reloaded"  "output"=" - Checking http://localhost:80 ...\n - Health check ok : 0 retry attempt(s).\n"
I0412 19:07:23.513991       1 router.go:618] template "msg"="router reloaded"  "output"=" - Checking http://localhost:80 ...\n - Health check ok : 0 retry attempt(s).\n"
I0412 19:13:26.571833       1 router.go:618] template "msg"="router reloaded"  "output"=" - Checking http://localhost:80 ...\n - Health check ok : 0 retry attempt(s).\n"
I0412 19:13:36.542543       1 router.go:618] template "msg"="router reloaded"  "output"=" - Checking http://localhost:80 ...\n - Health check ok : 0 retry attempt(s).\n"
```

Reply:
Hm, but I do see the following in the map file:
```
^test-route-with-path-with-ln-openshift-ingress-canary\.apps\.ci-ln-kliwjz2-72292\.origin-ci-int-gce\.dev\.rhcloud\.com\.?(:[0-9]+)?/foo
bar(/.*)?$ 0
bar(/.*)?$ 0with-path-with-cr-openshift-ingress-canary\.apps\.ci-ln-kliwjz2-72292\.origin-ci-int-gce\.dev\.rhcloud\.com\.?(:[0-9]+)?/foo
```
I'll add a check for `\n`.